<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
	<script src="https://unpkg.com/axios/dist/axios.min.js"></script>
  </head>
  <body>
    <div id="root"></div>
	<button id="btnAjax">Ajax</button>
	<button id="btnFecth">Fecth</button>
	<button id="btnAxios">Axios</button>
	<button id="btnPromise">Promise</button>
	<button id="btnAsyncFn">AsyncFn</button>
	<script>
	    btnAjax.onclick = function(){
            // Ajax 通过事件和回调函数实现异步
            var xhr = new XMLHttpRequest(); // 发起http请求，老式返回XML，现在json
	
	        // 监听 xhr 的 onload 事件
	        xhr.onload = function(){
	            //设置响应数据的处理方式
		        console.log(xhr.response)
	        };
	
	        // 设置请求信息
	        xhr.open("get", "https://raw.githubusercontent.com/mdn/learning-area/main/javascript/apis/fetching-data/can-store/products.json");
	
	        //发生请求
	        xhr.send();
        };
		
		btnFecth.onclick = function(){
            fetch("https://raw.githubusercontent.com/mdn/learning-area/main/javascript/apis/fetching-data/can-store/products.json")
			    .then(response => response.json())
                .then(data => console.log(data))
                .catch(() => console.log("Error"))				
        };
		
		btnAxios.onclick = function(){
		    axios
			
                .get("https://raw.githubusercontent.com/mdn/learning-area/main/javascript/apis/fetching-data/can-store/products.json")
			    .then(response => console.log(response))
                .catch(() => console.log("Error"))				
        };
		btnPromise.onclick = function(){
		    const promise = new Promise((resolve, reject)=>{
		        setTimeout(() =>{
				    let num = Math.ceil(Math.random()*10);
					if(num>=5){
					    resolve(num)
					} else {
					    reject("num<5")
					}
			    }, 1000)    
		    })
		    // 获取Promise的返回值，当Promise中的代码正常执行时，会自动通过then()方法的回调函数来返回结果
		    promise
			    .then(response => console.log(response))
				.catch((error) =>console.log("Error type: "+error))
		}
		
		//异步代码套壳成为异步函数
		
		function asyncFn(){
		    return new Promise((resolve, reject)=>{
			    setTimeout(()=>{
				    let num = Math.ceil(Math.random()*10)
					if(num>5){
					    resolve(5)
					} else{
					    reject("This's error,the number less than or equal to 5.")
					}
				})
			})
		}
		
		btnAsyncFn.onclick = function(){
		    asyncFn()
			    .then(response => console.log(response))
				//.then(console.log) //console.log作为异步的回调函数传入
				.catch((error) =>console.log(error))
		}
		
	    /*
		* Async and await
		*
		*
		*/
	</script>
  </body>
</html>
